### PR TITLE
BREAKING CHANGE: chore: sync with internal lint

### PR DESCRIPTION
**Pull request by ofrobots**
This synchronizes the lint rules to be closer to the lint rules used
internally at Google. Not all rules are available externally, so this
is not *exactly* the same.

Also updated the TypeScript version to 2.6.1.

A fair number of new lint rules are being applied now, which means that
the code will be alot more consistent with other Google code, but it
also means that users of gts may need to modify the source to pick up
this release. For most of the new lint rules, a fixer is also available
which should help with migration.

TypeScript 2.6.x enables contravariant types for function parameters,
which may also require some changes. See [1] for more info.

[1] https://blogs.msdn.microsoft.com/typescript/2017/10/31/announcing-typescript-2-6/

### CODE REVIEW

File: tslint.json [hunk 1]
```diff
@@ -1,31 +1,64 @@
...
   "rules": {
+    "array-type": [true, "array-simple"],
+    "arrow-return-shorthand": true,
+    "ban": [true,
+      {"name": "parseInt", "message": "tsstyle#type-coercion"},
+      {"name": "parseFloat", "message": "tsstyle#type-coercion"},
+      {"name": "Array", "message": "tsstyle#array-constructor"}
+    ],
+    "ban-types": [true,
+      ["Object", "Use {} instead."],
```
Comment (jinwoo) [inline tslint.json]: What's the rationale for this rule? Just curious.

File: tslint.json [hunk 2]
```diff
@@ -1,31 +1,64 @@
...
+    "use-isnan": true,
+    "variable-name": [
+      true,
+      "check-format",
+      "ban-keywords",
+      "allow-leading-underscore",
+      "allow-trailing-underscore"
+    ]
   }
 }
```
Comment (jinwoo) [inline tslint.json]: Nit. newline?

File: tslint.json (hunk 1 above)
Comment (kjin) [reply tslint.json]: +1, as a reader I prefer `Object`

Comment (ofrobots) [reply tslint.json]: https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html

Comment (ofrobots) [reply tslint.json]: More: https://stackoverflow.com/questions/18961203/typescript-any-vs-object/28795689#28795689

**Pull request merged**